File: Material Core/src/main/java/material/core/MaterialContextWrapper.java
package material.core;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.res.Resources;
import android.view.LayoutInflater;

public class MaterialContextWrapper extends ContextWrapper {
    private LayoutInflater inflater;
    private MaterialResources resources;

    public static Context wrap(Context context) {
        if (context instanceof MaterialContextWrapper || context instanceof MaterialContextThemeWrapper)
            return context;
        return new MaterialContextWrapper(context);
    }

    private MaterialContextWrapper(Context base) {
        super(base);
        resources = new MaterialResources(this, getAssets(), super.getResources().getDisplayMetrics(), super.getResources().getConfiguration());
    }

    @Override
    public Resources getResources() {
        return resources;
    }

    @Override
    public Object getSystemService(String name) {
        if (LAYOUT_INFLATER_SERVICE.equals(name)) {
            if (inflater == null) {
                inflater = LayoutInflater.from(getBaseContext()).cloneInContext(this);
            }
            return inflater;
        }
        return super.getSystemService(name);
    }
}

File: core/src/main/java/material/core/MaterialContextWrapper.kt
package material.core

import android.content.Context
import android.content.ContextWrapper
import android.content.res.Resources
import android.view.LayoutInflater

/**
 * Created by ali on 7/3/2018.
 */
class MaterialContextWrapper(base: Context?) : ContextWrapper(base) {
    private val resources = MaterialResources(this, assets, super.getResources().displayMetrics, super.getResources().configuration);
    private var inflater: MaterialLayoutInflater? = null

    companion object {
        @JvmStatic
        fun wrap(context: Context): Context = if (context is MaterialContextWrapper || context is MaterialContextThemeWrapper) context
        else MaterialContextWrapper(context)

    }

    override fun getResources(): Resources = resources
    override fun getSystemService(name: String?): Any {
        if (Context.LAYOUT_INFLATER_SERVICE == name) {
            if (inflater == null) inflater = MaterialLayoutInflater(LayoutInflater.from(baseContext),this)
            return inflater!!
        }
        else return super.getSystemService(name)
    }
}